xbox/xgu/xgu.h

```diff
@@ -0,0 +1,97 @@
+#include "nv2a_regs.h"
+#pragma GCC diagnostic pop
+
+#include <stdint.h>
+
+typedef enum {
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_UB_D3D     0
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_S1         1
+  XGU_FLOAT = NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_F,
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_UB_OGL     4
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_S32K       5
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_CMP        6
+} XguVertexArrayType;
+
+typedef enum {
+  XGU_END = NV097_SET_BEGIN_END_OP_END, // FIXME: Disallow this one?
+  XGU_POINTS = NV097_SET_BEGIN_END_OP_POINTS,
+  XGU_LINES = NV097_SET_BEGIN_END_OP_LINES,
+  XGU_LINE_LOOP = NV097_SET_BEGIN_END_OP_LINE_LOOP,
+  XGU_LINE_STRIP = NV097_SET_BEGIN_END_OP_LINE_STRIP,
+  XGU_TRIANGLES = NV097_SET_BEGIN_END_OP_TRIANGLES,
+  XGU_TRIANGLE_STRIP = NV097_SET_BEGIN_END_OP_TRIANGLE_STRIP,
+  XGU_TRIANGLE_FAN = NV097_SET_BEGIN_END_OP_TRIANGLE_FAN,
+  XGU_QUADS = NV097_SET_BEGIN_END_OP_QUADS,
+  XGU_QUAD_STRIP = NV097_SET_BEGIN_END_OP_QUAD_STRIP,
+  XGU_POLYGON = NV097_SET_BEGIN_END_OP_POLYGON
+} XguPrimitiveType;
+
+typedef enum {
+  XGU_VERTEX_ARRAY = NV2A_VERTEX_ATTR_POSITION,
+  XGU_NORMAL_ARRAY = NV2A_VERTEX_ATTR_NORMAL,
+  XGU_COLOR_ARRAY = NV2A_VERTEX_ATTR_DIFFUSE,
+  XGU_SECONDARY_COLOR_ARRAY = NV2A_VERTEX_ATTR_SPECULAR,
+  XGU_FOG_ARRAY = NV2A_VERTEX_ATTR_FOG,
+  XGU_TEXCOORD0_ARRAY = NV2A_VERTEX_ATTR_TEXTURE0,
+  XGU_TEXCOORD1_ARRAY = NV2A_VERTEX_ATTR_TEXTURE1,
+  XGU_TEXCOORD2_ARRAY = NV2A_VERTEX_ATTR_TEXTURE2,
+  XGU_TEXCOORD3_ARRAY = NV2A_VERTEX_ATTR_TEXTURE3
+} XguVertexArray;
+
+#define push_command(...) 0
+#define push_parameter(...) 0
+
+static inline DWORD* push_boolean(DWORD* p, bool enabled) {
+  return push_parameter(p, enabled ? 1 : 0);
+}
+
+static inline DWORD* push_command_boolean(DWORD* p, DWORD command, bool enabled) {
+  p = push_command(p, command, 1);
+  p = push_boolean(p, enabled);
+  return p;
+}
+
+static inline DWORD* push_command_parameter(DWORD* p, DWORD command, DWORD parameter) {
+  p = push_command(p, command, 1);
+  p = push_parameter(p, parameter);
+  return p;
+}
+
+static inline DWORD* xgu_begin(DWORD* p, XguPrimitiveType type) {
+
+  // Force people to use xgu_end instead
+  assert(type != NV097_SET_BEGIN_END_OP_END);
+
+  return push_command_parameter(p, NV097_SET_BEGIN_END, type);
+}
+
+static inline DWORD* xgu_end(DWORD* p) {
+  return push_command_parameter(p, NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END);
+}
+
+#define xgu_set_vertex_array_format(...) 0
+#define xgu_set_vertex_array_offset(...) 0
+
+static inline DWORD* xgu_elements16(DWORD* p, uint16_t* elements, unsigned int count) {
+  assert(count % 2 == 0);
+  // This one is a bit complicated
+  assert(false);
+  return p;
+}
+
+static inline DWORD* xgu_elements32(DWORD* p, uint32_t* elements, unsigned int count) {
+  // This one is a bit complicated
+  assert(false);
```

Comment:
See https://github.com/XboxDev/nxdk/blob/master/samples/mesh/main.c#L374-L378

This one isn't too complicated.